torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -646,8 +684,15 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
 
   // Get devices for tensors in the request message. This can throw if device
   // maps are not configured properly for this request.
-  auto devices =
-      getDevicesForTensors(clientPipe.pipe_->getRemoteName(), requestMessage);
+  std::vector<c10::DeviceIndex> devices;
+  if (deviceMap.empty()) {
+    devices =
+        getDevicesForRemote(clientPipe.pipe_->getRemoteName(), requestMessage);
+  } else {
+    // If deviceMap is specified, use that instead.
+    devices = getDevicesForTensors(
```

Comment:
Is it possible to consolidate `getDevicesForTensors` and `getDevicesForRemote` into one by letting it take an optional device map arg?

Reply:
`getDevicesForTensors` is actually a subset of `getDevicesForRemote`. `getDevicesForRemote` internally calls `getDevicesForTensors` to avoid any duplication in logic.

Reply:
Yep, I noticed that. What I didn't follow is why we need two similar but slightly different utility functions, and their didn't convey their difference? If all we need to letting the second utility function skip the map search and use the provided device map, this can potentially be done by using an optional arg? But this is a nit, please feel free to ignore.